### PR TITLE
Cosmos spark 3.4 support

### PR DESCRIPTION
**Pull request by TheovanKraay**
# Description

Updates required to support Spark 3.4 for Cosmos DB Spark Connector

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/pom.xml
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-4_2-12/pom.xml:180]: Has been updated to 2.14.2 in Spark 3.4

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/pom.xml
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-4_2-12/pom.xml:185]: Dito 2.14.2

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/README.md
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-4_2-12/README.md:36]: 4.19 has not been shipped for just Spark 3.3 - either it exisst across Spark 3.1, 3.2 and 3.3 or not yet at all

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/README.md
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-4_2-12/README.md:125]: 3-4

File: sdk/cosmos/spark.yml
Comment (benbp) [inline sdk/cosmos/spark.yml:55]: FYI I think the inner yaml file could be changed to something like:

```
stages:
  - ${{ each version in parameters.SparkVersions }}
     < rest of yaml template indented one more level here >
     < reference individual spark version like ${{ version }} >
```

Then you could just have this at the `spark.yml` level:

```
variables:
  - template: ../../eng/pipelines/templates/variables/globals.yml
  - name: AdditionalArgs
    value: ''

stages:
  - template: /sdk/cosmos/spark.databricks.yml
    parameters:
      SparkVersions:
        - 'azure-cosmos-spark_3-3_2-12'
        - 'azure-cosmos-spark_3-4_2-12'
```

I'm happy to help get this set up if interested.

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/CONTRIBUTING.md
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3-4_2-12/CONTRIBUTING.md:13]: nit: `3-3_2.12` -> `3-4_2.12`

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/README.md
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3-4_2-12/README.md:3]: nit: `Spark 3.3` -> `Spark 3.4`

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/pom.xml
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3-4_2-12/pom.xml:185]: Tried to update these but came up as banned dependencies, then adding to external dependencies broke things for other projects. Not sure how to add this only for our project without avoiding these errors?

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-4_2-12/pom.xml:185]: Dito - leave it as is for now.

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - the Jackson Databind mismatch is unfortunate - but unless Central team moves to 2.14 not really something we can change - so, let's see whether there are any problems first - but we'll need to keep an eye out on this in Databricks and Synapse (once a runtime exists there)

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/CHANGELOG.md
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-4_2-12/CHANGELOG.md:15]: Everything 4.20 and below can be removed from Spark 3.4 changelog?

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-4_2-12/CHANGELOG.md]: Fixed in directly inline

**Review by FabianMeiswinkel — CHANGES_REQUESTED** [review]
Please take a look at the Spark 3.4 CosmosRowConverter comment

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/README.md
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-4_2-12/README.md]: Resolved inline

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala:60]: I did this way because TimestampNTZType is a new data type, so of course it can't be added to base class at all (does not exist prior to 3.4). I have only overriden the methods that have match cases for each possible data type, in order to be able to add a case for the new data type (TimestampNTZType), though of course this does mean a lot of duplication: 

convertSparkDataTypeToJsonNodeConditionally 
convertSparkDataTypeToJsonNodeNonNull
convertToSparkDataType

The other methods are custom methods for the TimestampNTZType conversion. I did not see a more parsimonious way of doing it?

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala:63]: i Think this should be case TimestampNTZType if rowData.isInstanceOf[java.time.ocalDateTime] - to make sure if rowData is of a different type we get the catch all case exception instead of some invalid cast exception.

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala:125]: The Z suffix sounds wrong - it inidcates UTC time zone - the NTZ timestamp should not have any timezone suffix?

**Comment by FabianMeiswinkel** [comment]
Before merging please od one end-to-end validation using teh NYC Taxi-Data sample notebook. You might be able to discuss this with Annie - she is also doing some end-to-end validation - if we run that once on both of your changes we can save one of you some time

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala:63]: Ok, will add

**Review by FabianMeiswinkel — CHANGES_REQUESTED** [review]
Sorry, just blocking this PR to make sure it does not get merged before the fix we need to provide for ADF this week. We can merge this one immediately after the release happend

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM

**Pull request merged**
